lib: add option parameter for default value priority

Atemu opened the pull request:
It's sometimes the case that your option's default value should be additive such that setting it results in the default value being merged with whatever was set.

Previously, you had to add the default value in the implementation which might be many lines away from the option declaration and is generally an odd pattern.

This is a lot clearer, simpler and doesn't add much complexity.

If this becomes used more widely, it should be rendered in the generated options' manual and nixos-search.

## Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

Atemu commented:
Friendly ping.

roberth commented on lib/modules.nix:
```diff
@@ -780,9 +780,11 @@ let
-            { file = head opt.declarations; value = mkOptionDefault opt.default; }) ++ defs;
+      defs' = (optional (opt ? default) {
+        file = head opt.declarations;
+        value = (if opt.additive or false then id else mkOptionDefault) opt.default;
```
I don't think additive is quite the right term. Many types are mergeable, and `additive = false` would not negate that. That'd be unexpected.

I think we could perhaps simplify this to:

```suggestion
        value = mkOverride (opt.defaultPriority or defaultPriority) opt.default;
```

This simplifies the implementation a bit, uses the existing terminology (no conflicts there) and is a bit more powerful (for what it's worth).

Yet another possible syntax would be something like `mkOption { baseDefinition = v; }` (instead of `mkOption { default = v; additive = true; }`), but I think I prefer my prior suggestion.

infinisil replied:
`defaultPriority` is actually deprecated, and it's the default priority of `config` assignments, not `mkOptionDefault`. But I get the idea: Have the options specify a priority for the default.

I think neither the `additive` idea nor setting the priority works out in the end, because it breaks down as soon as you have a more deeply-nested value, e.g.

```nix
lib.mkOption {
  default = {
    foo = 0;
    bar.baz = 10;
    bar.qux = 20;
  };
```

While a user can change `foo` by itself, they can't change `bar.baz` without clearing `bar.qux` too.

What we really need is a default that sets it like
```nix
config = {
  foo = lib.mkOptionDefault 0;
  bar = {
    baz = lib.mkOptionDefault 10;
    qux = lib.mkOptionDefault 20;
  };
}
```

And for this I think an additional `lib.mkOption` attribute is justified, something like `recursiveDefault = true`. And it's not an existing module system concept.

Something very important that needs to be thought about though is allowing options to be backwards compatible, because if options all around just start adding `recursiveDefault = true`, it can break user configs.

infinisil replied:
I don't think the `additive` property works out, because it means that users will have to use `lib.mkForce` to change the defaults. What we really need is an option to push down `lib.mkOptionDefault` to the attribute leaves, e.g.

```nix
lib.mkOption {
  default = {
    foo = 0;
    bar.baz = 10;
    baz.qux = 20;
  };
  recursiveDefault = true;
}
```

turns into effectively

```nix
config = {
  foo = lib.mkOptionDefault 0;
  bar.baz = lib.mkOptionDefault 10;
  baz.qux = lib.mkOptionDefault 20;
}
```

This both overriding defaults without `lib.mkForce` and without clearing the rest.

And the `defaultPriority` idea either suffers from the same `lib.mkForce` problem if the priority is 100 (the default `config` priority) and the clearing problem for lower values (and the problem of default `config` values being ignored entirely for higher values).

---

Whatever solution we end up with though, we need to consider a way for options to stay backwards compatible. Because if they just add `recursiveDefault = true`, it can change behaviour.

Atemu replied:
> I don't think the `additive` property works out, because it means that users will have to use `lib.mkForce` to change the defaults.

Well yes, that is the intention ;)

This flag is the alternative for the pattern of setting the additive default in the implementation. You'd also need mkForce in that case.

This flag should only be used where the user could conceivably want the defaults to remain in place in all but extreme edge-cases.

> What we really need is an option to push down `lib.mkOptionDefault` to the attribute leaves

I think for that purpose it'd be preferable to do this:

```Nix
foo = mkOption {
  additive = true; # Or defaultPriority = ...
  default = {
    foo = lib.mkOptionDefault 0;
    bar.baz = lib.mkOptionDefault 10;
    baz.qux = lib.mkOptionDefault 20;
  };
};
```

This strikes me as a lot simpler than a new boolean flag because with the default being "normal" priority (what's the term here actually?), you have full control of the attrs' priorities and can simply do all your fine-grained priority declarations there. Before, any priorities here would have been clamped to the option default priority unconditionally.

roberth reviewed (commented):
I've put  my thoughts on the design in a comment.

I'll be interested to hear what @infinisil thinks about the design.

When that's agreed, we'll need:
- [x] tests in `lib/tests/modules.sh` 
- [x] documentation for the new `mkOption` arguments in the manual
- [ ] update the docs generation to show the new info
  - `nixos/lib/make-options-doc/default.nix`
  - `pkgs/tools/nix/nixos-render-docs/`

Atemu commented:
Thank you all for the feedback, I still intend to continue this but it may take a bit.

Atemu commented:
Another thing we may want to think about while fixing this issue is cases where you want to override the default value (i.e. remove an item from a list) but still allow other modules to merge in their values too.

Atemu committed:
lib/modules: refactor priorities

I needed to expose the baseline priority (and ideally the other priorities too)
for external use.

While I was at it, I gave mkDefault and defaultOverridePriority (not the same
thing in any way, confused yet?) better names.

Atemu committed:
lib: add option parameter for default value priority

When you define an option's default value, the expectation is that whenever the
option is set somewhere, the default value will be overridden by that setting
rather than merged.

It's sometimes the case however that your option's default value should be
merged with what is being set rather than overwritten by it; only allowing
explicit overrides via `mkForce`. Creating such an option was not possible
previously as all options' default values were forcefully clamped via
mkOptionDefault.

There were some workarounds such as creating additional options ("things" +
"extraThings") which you'd then have to explain to users and generally aren't
ergonomical to use.

Another workaround was to explicitly set options value to the default value in
the implementation but those two lines of code might be many lines apart from
another and it's generally an odd pattern. Example:

    {
      options.foo = lib.mkOption {
        default = [ "foo" ];
      };

      # Many more lines of code

      config.foo = options.foo.default;
    }

It is now possible to control this behaviour at option declaration which is much
clearer in the code and is introspect-able.

    {
      options.foo = lib.mkOption {
        default = [ "foo" ];
        defaultPriority = lib.modules.priorities.baseline;
      };

      # Many more lines of code
    }

The implementation is trivial.

If this becomes used more widely, it should be rendered in the generated options
manual and nixos-search.

Atemu commented:
I've changed the setting to take a priority value instead of a boolean and exposed the priority values in a sane manner.

Will take a look at docs and tests next.

Atemu committed:
lib/tests/modules: add tests for defaultPriority

Atemu committed:
nixos/doc: document mkDefault's defaultPriority

Atemu commented:
I've added documentation but I don't know how to touch nixos-render-docs. I think I'd prefer to not add it for now.

roberth commented on lib/modules.nix:
```diff
@@ -1023,12 +1025,32 @@ let
+
+  # Legacy aliases
+  mkDefault = mkConfigDefault;
+  defaultOverridePriority = priorities.baseline; # Deprecated! Please use priorities.baseline instead.
```
I don't think we're in a hurry, but let's plan the migration to start in July when the alternative is widely available.

```suggestion
  defaultOverridePriority = warnIf (oldestSupportedReleaseIsAtLeast 2505) "lib.modules.defaultOverridePriority is deprecated. Please use lib.modules.priorities.baseline instead." priorities.baseline;
```

Atemu replied:
Good idea, I'll mark both as deprecated as of next year and will check for usages inside of Nixpkgs.

roberth reviewed (commented):
Looks pretty good from a computational perspective, but this does create a problem with the meaning of the word "default" now, ambiguously referring to
1. certain priorities
2. a preset value having some declared priority that's not necessarily a "default" priority

We can handle a bit of ambiguity, but the possibility of having "a default that isn't a default" does kind of make it cross out of the grey area.

Ideally we'd rename `mkOption { default, defaultText }` to something like `mkOption { preset, presetText }` (and `presetPriority`). That'd be a huge migration, and canonicalizing the option declarations will have a measurable performance cost during such a migration.
Nonetheless, I do believe most modules are yet to be written...
We've managed the markdown migration pretty well, and this is a simpler syntax change with fewer steps (nothing analogous to _remove `mdDoc`_).

It's a heavy decision, but one or two such choices make the difference between devolving into a weird legacy system or something that's actually nice to use.

I think the migration path would be
1. now
 - add `preset` and `presetText` as alternatives
 - add `presetPriority`
 - set `default` and `defaultText` in `evalOptionValue` for compatibility with code that reads those values
2. after a release cycle
  - warn when any `default*` attributes is set, in `mkOption`
  - warn when `default*` is used, in `evalOptionValue`
3. after another release cycle
  - remove `default*`.

@Atemu deep apologies for the scope increase.
@infinisil am I going insane?

Atemu commented:
I can see where you're coming from but I think I still prefer `default` over `preset` actually; I think it's still a fitting name. It's still the default value if you don't touch it. This only changes what happens to that default value when you set the option and there it still remains the default value, just that it "adds" itself to whatever value is set.

I think it's perfectly expectable for configuration system's default values to remain in effect when an option is set, so I don't think this violates the principle of least surprise. I'd even argue it's more of a quirk of the NixOS module system that this wasn't possible before. We're used to only overrides of course but I don't think an outsider would consider it odd if both overriding and merging a default value was possible.

I don't see a dependency between PR and potentially renaming `default` though; it's entirely independent. We can just change the naming for defaultPriority later together with all the other default\* parameter names if/when we decide to do it. If such a change was already cut and dried, I'd wait of course but this appears to be the first time you bring it up and it's very much still up to debate whether it actually needs to happen.

I'd prefer to keep the name for `default` out of scope for this PR.

Atemu commented on lib/modules.nix:
```diff
@@ -1023,12 +1025,32 @@ let
+  priorities = {
+    # Priority of options' default values
+    optionDefault = 1500;
+    # Used in config sections of non-user modules to set a default.
```
I retained this docstring but it's terrible, it should be explained better.

Atemu commented on lib/modules.nix:
```diff
@@ -1023,12 +1025,32 @@ let
+    # Used in config sections of non-user modules to set a default.
+    # (This is not the default priority assigned to values!)
+    configDefault = 1000;
+    # The priority that values get assigned by default around which other
+    # priorities scale. This is the true "default" priority.
```
```suggestion
    # The priority that values get assigned by default around which other
    # priorities scale. This is the true "default" priority if you don't explicitly provide a priority in your config.
```

Atemu committed:
lib/modules: improve priorities doc comments

Atemu committed:
lib/modules: provide deprecation notices legacy priority aliases

It's a bit unfortunate to have to change things again but hopefully this
iteration will last for longer.

Atemu committed:
lib/modules: add mkBaseline

I don't expect this to be used much but it should exist for completeness' sake.

Atemu committed:
nixos/doc: rework options definitions priorities section

This makes the wording clearer with more examples and usage of the new names for
functions and priority values.

Atemu committed:
tree-wide: defaultOverridePriority -> priorities.baseline

Atemu committed:
lib/modules: keep mkDefault as an alias

Atemu commented:
Another point that just came to mind is that most of what this is useful for is already possible without this PR (as described in the commit message), it just isn't as nice to implement. It's actually already in use in a few places.
